### PR TITLE
CON-2138 - Initial implementation of the shopgate analytics plugin

### PR DESCRIPTION
**Pull request by reneeichhorn**

### CODE REVIEW

File: frontend/Plugin.js [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    });
+
+    this.registerEvents();
+  }
+
+  /**
+   * Registers all event handlers.
+   */
+  registerEvents() {
+    this.register.pageview((data, rawData) => {
```
Comment (Carsten89) [inline frontend/Plugin.js]: please add some links to the event spec

File: frontend/Plugin.js [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+          },
+        };
+      }
+
+      sgAnalytics('track', 'pageViewed', sdkData);
+      console.error('pageview', data, rawData, sdkData);
+    });
+
+    this.register.addToCart((data, rawData) => {
+      console.log(rawData);
```
Comment (Carsten89) [inline frontend/Plugin.js]: remove pls

File: frontend/Plugin.js [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      const sdkData = {
+        products: rawData.products.map(product => ({
+          number: product.uid || product.productNumber,
+          name: product.name,
+          currency: product.amount.currency,
+          price: product.amount.displayPrice ||
+           product.amount.grosss ||
+           product.amount.net,
+          priceStriked: product.amount.displayPriceStriked ||
```
Comment (Carsten89) [inline frontend/Plugin.js]: let's talk about the default value of priceStriked

File: frontend/Plugin.js [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        shipping: {
+          type: rawData.order.shipping.name,
+          price: rawData.order.shipping.amount.gross,
+        },
+        coupons: rawData.order.coupons.map(coupon => ({
+          code: coupon.code,
+          savings: coupon.amount.gross,
+          currency: coupon.amount.currency,
+        })),
+        products: rawData.order.products.map(product => ({
```
Comment (Carsten89) [inline frontend/Plugin.js]: this could be outsourced into a function and used in multiple places

File: frontend/Plugin.js [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+      };
+
+      sgAnalytics('track', 'productAddedToCart', sdkData);
+      console.error('addToCart', data, rawData, sdkData);
+    });
+
+    this.register.purchase((data, rawData) => {
+      console.error(data, rawData);
+      const sdkData = {
+        orderId: rawData.order.number,
```
Comment (Carsten89) [inline frontend/Plugin.js]: orderNumber?

File: frontend/Plugin.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+        totalPrice: rawData.order.amount.displayPrice ||
+          rawData.order.amount.gross ||
+          rawData.order.amount.net,
+        totalPriceStriked: rawData.order.amount.displayPriceStriked ||
+          rawData.order.amount.gross ||
+          rawData.order.amount.net,
+        currency: rawData.order.amount.currency,
+      };
+
+      console.error('checkoutCompleted', data, rawData, sdkData);
```
Comment (Carsten89) [inline frontend/Plugin.js]: remove pls

File: extension-config.json [hunk 7]
```diff
@@ -0,0 +1,20 @@
+{
+  "version": "1.0.0",
+  "id": "@shopgate/tracking-shopgate-analytics",
+  "components": [
+    {
+      "id": "init",
+      "path": "frontend/src/index.js",
```
[the remark this answers is not in the document]
Comment (reneeichhorn) [reply extension-config.json]: fixed

File: frontend/Plugin.js [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+   * @param {number} options.shopNumber The shop number.
+   * @param {number} [options.userId] The internal user id.
+   * @param {string} [options.pushToken] The push token of the device.
+   */
+  constructor({ stage, shopNumber, userId, pushToken }) {
+    super('sgAnalytics', { stage });
+
+    /* eslint-disable eslint-comments/no-unlimited-disable */
+    /* eslint-disable */
+    (function (src) { var s, r, t; r = false; s = document.createElement('script'); s.type = 'text/javascript'; s.src = src; t = document.getElementsByTagName('script')[0]; t.parentNode.insertBefore(s, t); window.__shopgate_aq = window.__shopgate_aq || []; window.sgAnalytics = function () { window.__shopgate_aq.push(arguments); }})
```
[the remark this answers is not in the document]
Comment (reneeichhorn) [reply frontend/Plugin.js]: done

File: frontend/Plugin.js [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+    (`path/to/tracking-sdk.${stage}.min.js`);
```
[the remark this answers is not in the document]
Comment (reneeichhorn) [reply frontend/Plugin.js]: sure

File: frontend/Plugin.js [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    /* eslint-enable */
+    /* eslint-enable eslint-comments/no-unlimited-disable */
+
+    sgAnalytics('setConfig', {
+      shopNumber,
+      pushToken,
+      sgUserId: userId,
+      channel: 'app',
```
[the remark this answers is not in the document]
Comment (reneeichhorn) [reply frontend/Plugin.js]: fixed

File: frontend/Plugin.js (hunk 1 above)
Comment (reneeichhorn) [reply frontend/Plugin.js]: done

File: frontend/Plugin.js [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    this.register.pageview((data, rawData) => {
+      let sdkData = {
+        url: data.page.merchantUrl,
+        type: rawData.page.name,
+      };
+
+      if (sdkData.type === 'productDetails' && rawData.product) {
+        sdkData = {
+          ...sdkData,
+          id: rawData.product.uid,
```
[the remark this answers is not in the document]
Comment (reneeichhorn) [reply frontend/Plugin.js]: done

File: frontend/Plugin.js [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+              rawData.product.amount.net,
+            identifiers: {
+              ean: rawData.product.identifiers.ean,
+            },
+          },
+        };
+      }
+
+      sgAnalytics('track', 'pageViewed', sdkData);
+      console.error('pageview', data, rawData, sdkData);
```
[the remark this answers is not in the document]
Comment (reneeichhorn) [reply frontend/Plugin.js]: removed

File: frontend/Plugin.js (hunk 2 above)
Comment (reneeichhorn) [reply frontend/Plugin.js]: removed

File: frontend/Plugin.js (hunk 3 above)
Comment (reneeichhorn) [reply frontend/Plugin.js]: as discussed

File: frontend/Plugin.js [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+            product.amount.net,
+          identifiers: {
+            ean: product.identifiers.ean,
+          },
+          quantity: product.quantity,
+        })),
+      };
+
+      sgAnalytics('track', 'productAddedToCart', sdkData);
+      console.error('addToCart', data, rawData, sdkData);
```
[the remark this answers is not in the document]
Comment (reneeichhorn) [reply frontend/Plugin.js]: done

File: frontend/Plugin.js (hunk 5 above)
Comment (reneeichhorn) [reply frontend/Plugin.js]: changed

File: frontend/Plugin.js (hunk 4 above)
Comment (reneeichhorn) [reply frontend/Plugin.js]: done

File: frontend/Plugin.js (hunk 6 above)
Comment (reneeichhorn) [reply frontend/Plugin.js]: done

File: package-lock.json [hunk 14]
```diff
@@ -0,0 +1,1832 @@
+{
```
[the remark this answers is not in the document]
Comment (reneeichhorn) [reply package-lock.json]: done

File: frontend/index.js [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+
+import Plugin from './Plugin';
+import config from '../config';
+
+/**
+* Read the config and create the tracking plugin.
+* @return {Object}
+*/
+export default function init() {
+  return new Plugin(config);
```
[the remark this answers is not in the document]
Comment (reneeichhorn) [reply frontend/index.js]: As discussed, everything expect the `stage` was removed from the config. We need to add the remaining data to the tracking initialization. 
Follow up story: CON-2464

File: frontend/tests/Plugin.js [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+    /* eslint-disable no-new */
+    new Plugin({ stage: 'development', shopNumber: '1234', access: 'App' });
+    /* eslint-enable no-new */
+
+    SgTrackingCore.registerFinished();
+  });
+
+  it('should register specified events with core', () => {
+    expectedSubscriptions.forEach((eventName) => {
+      expect(SgTrackingCore.register[eventName].called).equals(true);
```
Comment (Carsten89) [inline frontend/tests/Plugin.js]: expect(SgTrackingCore.register[eventName]).to.have.been.calledOnce

**Pull request merged**